lib/PuppeteerSharp/NetworkManager.cs

```diff
@@ -247,53 +242,44 @@ private async Task OnRequestInterceptedAsync(RequestInterceptedResponse e)
 
             var requestHash = e.Request.Hash;
             var requestId = _requestHashToRequestIds.FirstValue(requestHash);
-            if (requestId != null)
+            if (requestId != null && _requests.TryGetValue(requestId, out var request))
             {
-                _requestIdToRequestWillBeSentEvent.TryGetValue(requestId, out var requestWillBeSentEvent);
-
-                if (requestWillBeSentEvent != null)
+                if (request.RequestWillBeSentEvent != null)
                 {
-                    await OnRequestAsync(requestWillBeSentEvent, e.InterceptionId);
+                    await OnRequestAsync(request, e.InterceptionId);
                     _requestHashToRequestIds.Delete(requestHash, requestId);
-                    _requestIdToRequestWillBeSentEvent.Remove(requestId);
                 }
             }
-            else
-            {
-                _requestHashToInterceptionIds.Add(requestHash, e.InterceptionId);
-            }
         }
 
         private async Task OnRequestAsync(RequestWillBeSentPayload e, string interceptionId)
         {
-            Request request;
             var redirectChain = new List<Request>();
             if (e.RedirectResponse != null)
             {
-                _requestIdToRequest.TryGetValue(e.RequestId, out request);
                 // If we connect late to the target, we could have missed the requestWillBeSent event.
-                if (request != null)
+                if (_requests.TryGetValue(e.RequestId, out var redirectRequest))
                 {
-                    HandleRequestRedirect(request, e.RedirectResponse);
-                    redirectChain = request.RedirectChainList;
+                    HandleRequestRedirect(redirectRequest, e.RedirectResponse);
+                    redirectChain = redirectRequest.RawRequest.RedirectChainList;
                 }
             }
-            if (!_requestIdToRequest.TryGetValue(e.RequestId, out var currentRequest) ||
-              currentRequest.Frame == null)
+            if (!_requests.TryGetValue(e.RequestId, out var request) ||
+              request.RawRequest.Frame == null)
             {
                 var frame = await FrameManager?.GetFrameAsync(e.FrameId);
 
-                request = new Request(
+                request.RawRequest = new Request(
```

Comment:
WorkerTests.ShouldEmitCreatedAndDestroyedEvents is failing around here. `request` is null, so we fail setting `RawRequest`